ambry-router/src/main/java/com/github/ambry/router/CompressionService.java

```diff
@@ -0,0 +1,350 @@
+   * For example, if the compressed data size is smaller than original data by certain % (exceeds threshold),
+   * the compressed data is accepted and return compressed data; otherwise, compression is discarded.
+   * This method emits metrics along the process.  It returns null if failed instead of throwing exceptions.
+   * This method is typically called for the first chunk to determine whether to compress the entire blob or not.
+   * Subsequent data chunks are compressed using compressChunk().
+   *
+   * @param chunkBuffer The PutChunk buffer to compress.
+   * @param isFullChunk Whether this is a full size chunk (4MB) or smaller.
+   * @return Returns a new compressed buffer.  It returns null is no compression is applied.
+   */
+  public ByteBuf compressChunk(ByteBuf chunkBuffer, boolean isFullChunk) {
+    Objects.requireNonNull(chunkBuffer, "blobProperties cannot be null.");
+
+    // Check the blob size.  If it's too small, do not compress.
+    int sourceDataSize = chunkBuffer.readableBytes();
+    if (sourceDataSize < compressionConfig.minimalSourceDataSizeInBytes) {
+      logger.info("No compression applied because source data size " + sourceDataSize + " is smaller than minimal size "
+          + compressionConfig.minimalSourceDataSizeInBytes);
+      // Emit metrics to count how many compressions skipped due to small size.
+      compressionMetrics.compressSkipRate.mark();
+      compressionMetrics.compressSkipSizeTooSmall.inc();
+      return null;
+    }
+
+    // Convert ByteBuf to byte[] before applying compression.
+    byte[] sourceBuffer;
+    int sourceBufferOffset;
+    try {
+      Triple<byte[], Integer, Integer> bufferInfo = convertByteBufToByteArray(chunkBuffer);
+      sourceBuffer = bufferInfo.getLeft();
+      sourceBufferOffset = bufferInfo.getMiddle();
+      sourceDataSize = bufferInfo.getRight();
+    } catch (Exception ex) {
+      logger.error(String.format("Cannot compress source data because failed to convert ByteBuf to byte[]."
+              + "chunkBuffer.hasArray() = %s, chunkBuffer.readableBytes = %d, chunkBuffer.arrayOffset = %d",
+          chunkBuffer.hasArray(), chunkBuffer.readableBytes(), chunkBuffer.arrayOffset()), ex);
+
+      // Emit metrics for count how many compressions skipped due to buffer conversion error.
+      compressionMetrics.compressErrorRate.mark();
+      compressionMetrics.compressErrorBufferConversion.inc();
+      return null;
+    }
+
+    // Get default compressor.  If config missing, fallback to Zstandard.
+    String algorithmName = compressionConfig.algorithmName;
+    Compression defaultCompressor = allCompressions.get(algorithmName);
+    if (defaultCompressor == null) {
+      // Emit metrics for default compressor name does not exist.
+      compressionMetrics.compressErrorConfigInvalidCompressorName.inc();
+      logger.error("The default compressor algorithm, " + algorithmName
+          + ", specified in config does not exist.  This config is ignored.");
```

Comment:
if the default compression doesn't exist in the CompressionMap, then it will never exist unless we change the config or code and release a new version. If this is the case, then we should do this in constructor so we don't see this error message over and over again.

Reply:
This log should appear only once because the code after this log sets the default to Zstd.  In my opinion, it is not a good practice to put logic to constructor.   Constructor is intended for initialization and parameter checking, not to perform any real action/work.  Actions should be in methods.